R/resolver.R

```diff
@@ -0,0 +1,373 @@
+#   }
+#   if (ncol(new_data) <= 0L) {
+#     stop("Can't pull variable: No variable is available.")
+#   }
+#   type <- determine_helper(type, colnames(new_data), new_data)
+#
+#   # Not possible to know what is happening
+#   if (is.delayed(type)) {
+#     return(list(type = type, data = NULL))
+#   }
+#
+#   if (length(type$select) > 1) {
+#     list(type = type, data = data[type$select])
+#   } else {
+#     list(type = type, data = data[[type$select]])
+#   }
+# }
+
+# @export
+# determine.mae_experiments <- function(type, data, ...) {
+#   if (!requireNamespace("MultiAssayExperiment", quietly = TRUE)) {
+#     stop("Requires 'MultiAssayExperiment' package.")
+#   }
+#   new_data <- experiments(data)
+#   type <- determine_helper(type, names(new_data), new_data)
+#
+#   # Not possible to know what is happening
+#   if (is.delayed(type)) {
+#   }
+#
+#   if (!is.delayed(type) && length(type$select) > 1) {
+#     list(type = type, data = new_data[type$select])
+#   } else if (!is.delayed(type) && length(type$select) == 1) {
+#     list(type = type, data = new_data[[type$select]])
+#   } else {
+#     return(list(type = type, data = NULL))
+#   }
+# }
+
+# @export
+# determine.mae_sampleMap <- function(type, data, ...) {
+#   if (!requireNamespace("MultiAssayExperiment", quietly = TRUE)) {
+#     stop("Requires 'MultiAssayExperiment' package.")
+#   }
+#
+#   new_data <- sampleMap(data)
+#   type <- determine_helper(type, names(new_data), new_data)
+#
+#   # Not possible to know what is happening
+#   if (is.delayed(type)) {
+#     return(list(type = type, data = NULL))
+#   }
+#
+#   if (length(type$select) > 1) {
+#     list(type = type, data = data[type$select])
+#   } else {
+#     list(type = type, data = data[[type$select]])
+#   }
+# }
+
+#' @export
+determine.values <- function(type, data, ...) {
+  if (!is.numeric(data)) {
+    d <- data
+    names(d) <- data
+  } else {
+    d <- data
+  }
+  sel <- selector(d, type$names)
+  type$names <- data[sel]
+
+
+  sel2 <- selector(d[sel], type$select)
+  type$select <- data[sel][sel2]
+
+  # Not possible to know what is happening
+  if (is.delayed(type)) {
+    return(list(type = type, data = NULL))
+  }
+
+  list(type = type, data = data[sel])
+}
+
+orig <- function(x) {
+  attr(x, "original")
+}
+
+unorig <- function(x) {
+  attr(x, "original") <- NULL
+  x
+}
+
+eval_type_select <- function(type, data) {
+  stopifnot(is.character(type$names))
+
+  orig_select <- orig(type$select)
+  if (length(orig_select) == 1L) {
+    orig_select <- orig_select[[1L]]
+  }
+
+  names <- seq_along(type$names)
+  names(names) <- type$names
+  select_data <- selector(data, type$select)
+
+  # Keep only those that were already selected
+  new_select <- intersect(unique(names(c(select_data))), unorig(type$names))
+  if (is.null(new_select)) {
+    stop("No ", is(type), " selection is possible.")
+  }
+  attr(new_select, "original") <- orig_select
+
+  type$select <- new_select
+
+  type
```

Comment:
`select` shouldn't be a "function" of data but of choices. What if we reduce the complexity to this (plus some additional warnings/assertions within this if):

```suggestion
eval_type_select <- function(type) {
  stopifnot(is.character(type$names))

  orig_select <- orig(type$select)
  new_select <- if (!checkmate::test_integerish(orig_select)) {
    type$names[orig_select]
  } else {
    intersect(orig_select, type$names)
  }
  attr(new_select, "original") <- orig_select
  type$select <- new_select

  type
```

Reply:
@llrs-roche and @pawelru please convince me why this should work. And explain what you'd do if `intersect(select, x)`  results in `character(0)`.

```r
variables(
  x = where(is.character), # possible choices
  select = where(function(x){n_distinct(x) > 5}) # initial selection
 )
```

Reply:
> please convince me why this should work. And explain what you'd do if `intersect(select, x)` results in `character(0)`.

Mmh, this could either be an error or just leave the user to pick one of the choices. 

> Not only by name. IMO select should be a "function" of choices. 

Ok, butAnd then we need the tidyselect help to evaluate the function, which is what I tried to do here.

Reply:
> It results in app-user facing punished with error made by an app-developer lost in too complex api.

Maybe the app-developer would see the error before sharing/deploying it. But it's true we shouldn't try to guide them better than this.

> Empty initial selection could be also an error (depending on module's validation).

This could be handled at the module level instead of at the extraction level.

Reply:
> Maybe the app-developer would see the error before sharing/deploying it. But it's true we shouldn't try to guide them better than this.

No easy to address. This is run and resolved inside of the teal_module's server so it takes a while before extract meets data.

> This could be handled at the module level instead of at the extraction level.

Please elaborate more

Reply:
> > Maybe the app-developer would see the error before sharing/deploying it. But it's true we shouldn't try to guide them better than this.
> 
> No easy to address. This is run and resolved inside of the teal_module's server so it takes a while before extract meets data.

Indeed

> 
> > This could be handled at the module level instead of at the extraction level.
> 
> Please elaborate more

Some modules will only allow to select one while others will allow to select multiple variables. So I think this kind of validation shouldn't be on the extraction feature. Between not selecting something and selecting something is more debatable in my opinion. But if we already need to include validation for one or more than one selection we can add validation between 0 and 1 at the same time.